src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs

```diff
@@ -366,32 +361,16 @@ private void OnDependenciesSnapshotChanged(object sender, SnapshotChangedEventAr
 
             lock (_treeUpdateLock)
             {
-                if (_treeUpdateQueueTask == null || _treeUpdateQueueTask.IsCompleted)
-                {
-                    _treeUpdateQueueTask = ThreadingService.JoinableTaskFactory.RunAsync(async () =>
-                    {
-                        if (TasksService.UnloadCancellationToken.IsCancellationRequested)
-                        {
-                            return;
-                        }
-
-                        await BuildTreeForSnapshotAsync(snapshot).ConfigureAwait(false);
-                    }).Task;
-                }
-                else
-                {
-                    _treeUpdateQueueTask = _treeUpdateQueueTask.ContinueWith(
-                        t => BuildTreeForSnapshotAsync(snapshot), TaskScheduler.Default);
-                }
+                BuildTreeForSnapshot(snapshot);
```

Comment:
>BuildTreeForSnapshot [](start = 16, length = 20)

Did you realize what old code was doing? It was ensuring that all snapshot updates were applied to the tree in the order they were received.. Since snapshot is updated by changes in multiple DT builds associated with unresolved and then resolved rules for dependencies. Since everything is async older snapshots could be applied to the tree before later ones potentially, which could result in weird tree state.

Reply:
If implementation could be changed, but the overall behavior should remain the same , i would not do this change.

---
In reply to: [171345584](https://github.com/dotnet/project-system/pull/3326#discussion_r171345584) [](ancestors = 171345584)

Reply:
I chatted with Lifeng offline and we agree that queue logic is needed there. The only thing that might be changed if desired is how queue is implemented - avoid async tasks there if possible. I will leave it to you to decide if you want to change that.

---
In reply to: [171345891](https://github.com/dotnet/project-system/pull/3326#discussion_r171345891) [](ancestors = 171345891,171345584)